### ISSUE TITLE
Application really slow when using APM Cloud

### ISSUE THREAD

**Issue by amcsi** [nipwaayoni/elastic-apm-laravel#27]
A webpage of ours would normally take 3 seconds to load, but with elastic-apm-laravel enabled it gets increased to 40 seconds :exclamation: 

It would be great to improve performance by either making sure the request would be async, or by making sure the requests are sent to apm-server only after the current request to our applications have been responded to.

**Comment by dstepe** [comment]
I've not used the APM cloud offering. I've only used local instances and have not observed this type of impact. Can you use `curl` or another alternate client to rule out a connectivity issue outside of PHP? I'm not sure what could be within the PHP package that could cause this behavior.

While an async request for this would make a lot of sense, I'm not aware of PHP support for a true `async` request like you have in JavaScript. I hope they find a way to address this in the official client under development now.

**Comment by amcsi** [comment]
@dstepe by async I mean dispatch the request, but _not_ wait for the response. There should be ways to do that in PHP resulting in the call to make the request return control to the PHP flow immediately.

the big problem is that before the user's request is responded to, they have to wait for the backend to send all the data to elastic-apm and wait for the response back. So especially if there's a connectivity issue like you say, it's bad that one has to wait for the bad connectivity to finish before getting a response.

**Comment by amcsi** [comment]
For async requests, you may need to look into using `\Http\Client\HttpAsyncClient::sendAsyncRequest()` rather than `\Psr\Http\Client\ClientInterface::sendRequest()`.

Oh, also maybe the problem simply is that you are calling `$this->agent->send();` at the end of `handle()` in the middleware. That happens before a response is sent to the client making a request to the Laravel app. I do also see `$this->agent->send();` called within `terminate()` which I believe is run after the Laravel request has been responded to. Maybe you could remove the one that's in `handle()` and just rely on the one in `terminate()` running?

**Comment by amcsi** [comment]
My colleague seems to have gotten the performance up by changing what was in `handle()`, `$this->agent->send();` to:

```php
App::terminating(function () use ($request) {
    $this->agent->send();
});
```

He couldn't even rely on the `terminate()` method, because apparently that only works on FastCGI, whereas the php7.4-apache Docker image he uses for local development uses prefork.

**Comment by dstepe** [comment]
That was my interpretation of the `handle` method as well, and why I've actually removed in the current development branch. It was actually causing my Docker image to call `send` twice since I'm using php-fpm.

I looked at the async request option not long ago. I did not pursue it at the time because the change to no longer package an HTTP client dependency in the project, we can't guarantee the user provided client would support the behavior. I would like to fix that in the underlying Agent package and will look at it again.

Using `App::terminating` in the Laravel package is a nice alternative. I'll look at that as well.

**Comment by amcsi** [comment]
Maybe you could check if `$client instanceof HttpAsyncClient` or `if (function_exists([$client, 'sendAsyncRequest'])`, you could conditionally do it the async way?

**Comment by dstepe** [comment]
I just published v7.2.0 of this package with some changes that might be helpful. I have opened an issue in the Agent package to look into the async request. 

The severe performance issue you observed seems like more than just an issue with the Agent and these "fixes" are just working around the problem. You mentioned APM Cloud, can you try with any other APM service to compare? What HTTP client implementation are you using?

I added pre/post commit hooks to the Agent package which you can use to log the request/response for APM data. You can do something like this in your `AppServiceProvider`:

```php
$this->app->bind(AgentBuilder::class, function () {
    $builder = new AgentBuilder();

    $builder->withPreCommitCallback(function (RequestInterface $request) {
        Log::info(sprintf('Pre commit url is: %s', $request->getUri()));
    });

    $builder->withPostCommitCallback(function (ResponseInterface $response) {
        Log::info(sprintf('Post commit response status: %s', $response->getStatusCode()));
        Log::debug($response->getBody()->getContents());
    });

    return $builder;
});
```

This would allow you to examine the request/response more closely to see if there is a clue in there.

**Comment by veelasky** [comment]
how about sending it over queue?! and let the queue driver (or horizon) handle the rest. that way it still can handle `send()` method without blocking any underlying response.

**Comment by dstepe** [comment]
While some form of true asynchronous sending of the data at the agent level would be optimal, there are two issues. First, do your point about queues, that agent package is not specifically a Laravel implementation and must accommodate uses cases not involving Laravel. The agent would need to somehow implement (or provide a means to implement) the queue functionality. Second, PHP's support for asynchronous events is limited. I found that trying to implement that in the agent was problematic due to how the curl extension's event loop worked.

The current "best" solution is to defer sending the APM data in the agent until after Laravel has sent the response to the caller. The transaction should appear "complete" to the end user at that point, but PHP can continue processing the APM data.

I suggest looking at an alternate Laravel package: https://github.com/arkaitzgarro/elastic-apm-laravel

I have started working with them to update their package to the newest agent as well as enhance the functionality of both. I expect to deprecate this Laravel package once there is an official release. It makes a lot of sense to collaborate rather than compete.
